lib/using_restricted_array.rb

```diff
@@ -1,61 +1,130 @@
+# Quadratic (multiply when loops are nested inside of one another; O(n*n) or O(n^3))
+
+# SPACE COMPLEXITY:
+# Contant: O(1)
+# Linear: O(n) <-- amount of space taken to loop through something takes the same amount of space (i takes 2 bytes to hold the counter, max (variable) takes 2 bites to hold the current max value, etc.)
+# Quadratic: O(n*m) duel btwn jedis (m) and siths (n), a winner is determined
+# -----
+
 # Calculates the length of the restricted array. All values are integers.
 # The restricted_array is terminated by 'nil' i.e. array[length] = nil
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) where n is the number of values in the array
+# Space complexity: O(1) constant because i always takes up the same amount of space; the method is not dependant upon the length of the
+
 def length(array)
-  raise NotImplementedError
+  i = 0
+  until array[i] == nil
+    # also valid --> while array[i] != nil
+    i += 1
+  end
+  return i
 end
 
-# Prints each integer values in the array
-# Time complexity: ?
-# Space complexity: ?
+# Prints each integer value in the array
+# Time complexity: O(n) where n is the number of values in the array
+# Space complexity: O(1)
 def print_array(array)
-  raise NotImplementedError
+  length = length(array)
+  i = 0
+  while i < length
+    print array[i]
+    print " "
+    i += 1
+  end
 end
 
 # For an unsorted array, searches for 'value_to_find'.
 # Returns true if found, false otherwise.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) where n is the length of the array
+# Space complexity: O(1)
 def search(array, length, value_to_find)
-  raise NotImplementedError
+  length.times do |i|
+    return true if array[i] == value_to_find
+  end
+  return false
 end
 
 # Finds and returns the largest integer value the array
 # Assumes that the array is not sorted.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) where n is the length of the array
+# Space complexity: O(1)
 def find_largest(array, length)
-  raise NotImplementedError
+  largest_value = array[0]
+  i = 0
+  length.times do |item|
+    (largest_value = array[i]) if (array[i] > largest_value)
+    i += 1
+  end
+  return largest_value
 end
 
 # Finds and returns the smallest integer value in the array
 # Assumes that the array is not sorted.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) where n is the length of the array
+# Space complexity: O(1)
 def find_smallest(array, length)
-  raise NotImplementedError
+  smallest_value = array[0]
+  i = 0
+  length.times do
+    (smallest_value = array[i]) if (array[i] < smallest_value)
+    i += 1
+  end
+  return smallest_value
 end
 
 # Reverses the values in the integer array in place
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) where n is the length of the array
+# Space complexity: O(1)
 def reverse(array, length)
-  raise NotImplementedError
+  i = 0
```

Comment:
(Personal preference) nit: use more descriptive names than `i` and `j`?

(Aside: some programmers prefer shorter names for loop iterators, while others prefer more explanatory ones - so this is mostly a matter of your/your team's personal preferences.)